Pluggable browser providers (#385)
scrapy-plugins/scrapy-playwright: docs/pluggable-browser-providers.md

# Pluggable browser providers

Third-party projects such as [patchright](https://pypi.org/project/patchright/),
[camoufox](https://pypi.org/project/camoufox/), and
[invisible_playwright](https://github.com/feder-cr/invisible_playwright) provide
drop-in replacements for Playwright's browser startup while keeping standard
`Browser`/`BrowserContext`/`Page` objects. Because those objects are unchanged,
pages, contexts and routing code keeps working — only how the browser is
started and stopped differs.

scrapy-playwright exposes a single extension point for this: the
`PLAYWRIGHT_BROWSER_PROVIDER` setting. It takes the import path of a *browser
provider* — a class that owns the browser lifecycle. The built-in provider wraps
vanilla Playwright; you can integrate any other backend by pointing the setting
at your own provider class.

## The `PLAYWRIGHT_BROWSER_PROVIDER` setting

Type `str` or `type`, default `"scrapy_playwright.provider.PlaywrightBrowserProvider"`.

```python
# settings.py
PLAYWRIGHT_BROWSER_PROVIDER = "myproject.providers.CustomBrowserProvider"
```

The value is resolved with
[`scrapy.utils.misc.load_object`](https://docs.scrapy.org/en/latest/topics/api.html),

> so it may be either an import path string or the provider class directly. The

Reviewer: ```suggestion
The value may be either an import path string or the provider class directly. The
```

The rest is an implementation detail 🙂

class is instantiated with a configuration object (see
[Reading configuration](#reading-configuration)) and drives every browser that
scrapy-playwright uses. When the setting is not set, the built-in
`PlaywrightBrowserProvider` is used.

## The `BrowserProvider` interface

A provider is a class that accepts a configuration object in its constructor and

> implements the following asynchronous lifecycle. You can subclass

Reviewer: ~~I think protocol classes are not meant to be subclassed.~~ And I was wrong. Subclassing is OK, and even recommended by some.

`scrapy_playwright.provider.BrowserProvider` (a `typing.Protocol`) to document
the intent, but it is not required — any class with these methods works.

```python
from playwright.async_api import Browser, BrowserContext
from scrapy.exceptions import NotSupported

class BrowserProvider:

>     def __init__(self, config) -> None:

Reviewer: Since you are typing, you may as well type config here.

license for authoritative, up-to-date usage, and evaluate any third-party
dependency yourself before using it.

## Example: patchright

Patchright mirrors the Playwright async API (Chromium only), so a provider can
use it exactly like Playwright, including persistent contexts.

```python
from contextlib import AsyncExitStack

class PatchrightBrowserProvider:
    def __init__(self, config):
        self.config = config
        self.stack = AsyncExitStack()
        self.browser_type = None

    async def start(self):
        from patchright.async_api import async_playwright

        playwright = await self.stack.enter_async_context(async_playwright())
        self.browser_type = getattr(playwright, self.config.browser_type_name)

    async def launch_browser(self):
        return await self.browser_type.launch(**self.config.launch_options)

    async def launch_persistent_context(self, context_kwargs):
        return await self.browser_type.launch_persistent_context(**context_kwargs)

    async def close(self):
        await self.stack.aclose()
```

```python
# settings
PLAYWRIGHT_BROWSER_PROVIDER = "myproject.providers.PatchrightBrowserProvider"

> PLAYWRIGHT_BROWSER_TYPE = "chromium"

Reviewer: I wonder if we could make it so that a provider can make this setting unnecessary.

Author: I don't think so, unfortunately. `Browser` instances expose this via `browser.browser_type.name`, but persistent contexts do not have a linked browser type. The browser name is used internally in the handler, and there wouldn't be a way to retrieve it in persistent contexts.
An alternative would be to require browser providers to declare a browser name attribute, but that would be basically the same as the current setting.
